### PR TITLE
Feature/universal types

### PR DESCRIPTION
**Pull request by wer1st**
### Description of the Change

- Add Universal type for types are different from actor version to actor version.
- Update Claims with new approach.
- Update MinerInfo with new approach.
- Delete StateProvider.
- Delete StateManager.

### Benefits

Now we can use:

- `Universal<type>`
- `CbCidT<Universal<type>>`
- `adt::Array<Universal<type>>`

as fields of any CBORed structures.

### CODE REVIEW

File: core/primitives/cid/cid.hpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ namespace fc {
     explicit CID(const CbCid &cid);
     CID(const ActorCodeCid &cid);
 
-    ~CID() = default;
+    virtual ~CID() = default;
```
Comment (turuslan) [inline core/primitives/cid/cid.hpp:51]: please remove `virtual` here.
also may remove default destructor

File: core/vm/actor/version.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <type_traits>
```
Comment (turuslan) [inline core/vm/actor/version.hpp:8]: please remove, WithActorVersion subclass not checked anymore

File: core/vm/actor/builtin/types/type_manager/universal.hpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "common/outcome.hpp"
+#include "storage/ipfs/datastore.hpp"
```
Comment (turuslan) [inline core/vm/actor/builtin/types/type_manager/universal.hpp:9]:
```suggestion
#include "cbor_blake/ipld_cbor.hpp"
```

File: core/vm/actor/version.hpp (hunk 2 above)
Comment (wer1st) [reply core/vm/actor/version.hpp:8]: Missed it, thanks

File: core/vm/actor/version.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    kVersion3 = 3,
+    kVersion4 = 4,
+    kVersion5 = 5,
+  };
+
+  struct WithActorVersion {
+    ActorVersion actor_version{};
+
+    WithActorVersion() = default;
+    WithActorVersion(ActorVersion v) : actor_version{v} {}
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/version.hpp:23]:
```suggestion
    explicit WithActorVersion(ActorVersion v) : actor_version{v} {}
```

File: core/vm/actor/version.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    operator ActorVersion() const {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/version.hpp:25]:
```suggestion
    explicit operator ActorVersion() const {
```

File: core/vm/actor/version.hpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace fc::vm::actor {
+  enum class ActorVersion {
+    kVersion0 = 0,
+    kVersion2 = 2,
+    kVersion3 = 3,
+    kVersion4 = 4,
+    kVersion5 = 5,
+  };
+
+  struct WithActorVersion {
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/version.hpp:19]: I suggest add protected destructor, so it is impossible to delete derived object via base class. I urge to follow [guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#discussion-make-base-class-destructors-public-and-virtual-or-protected-and-non-virtual).

Comment (turuslan) [reply core/vm/actor/version.hpp:19]: No

File: core/vm/actor/version.hpp (hunk 5 above)
Comment (turuslan) [reply core/vm/actor/version.hpp:25]: Isn't it redundant here?

Comment (Alexey-N-Chernyshov) [reply core/vm/actor/version.hpp:25]: Conversion function is not used and may be removed.

**Review by turuslan — APPROVED** [review]

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**